### DISCUSSION TITLE
How are function attributes like `__pure` determined?

### DISCUSSION THREAD

**Discussion by jryans** [Vector35/binaryninja-api#5872] [Q&A] [answered]
I've noticed that Binary Ninja can label functions with `__pure`, and I see that the docs [explain](https://docs.binary.ninja/guide/types/attributes.html#pure-functions) the precise meaning used for that term.

How is Binary Ninja determining attributes like this? Is it performing its own dataflow analysis, or is it consulting some pre-computed source of information?

**Comment by plafosse** [comment, marked as answer]
We're performing our own analysis of these functions based up on the IL to determine this.
